File: cpp/tests/BinaryTreeZigzagLevelOrderTraversalTest.cpp
#include "BinaryTreeZigzagLevelOrderTraversal.h"
#include "gtest/gtest.h"

using namespace lcpp;

TEST(BinaryTreeZigzagLevelOrderTraversal, Solution103_1) {
  auto S103_1 = Solution103_1();

  auto Root1 = newTree({3, 9, 20, NTNode, NTNode, 15, 7});
  std::vector<std::vector<int>> const Expected1{{3}, {20, 9}, {15, 7}};
  EXPECT_EQ(Expected1, S103_1.zigzagLevelOrder(Root1));
  deleteTree(Root1);

  auto Root2 = newTree({1});
  std::vector<std::vector<int>> const Expected2{{1}};
  EXPECT_EQ(Expected2, S103_1.zigzagLevelOrder(Root2));
  deleteTree(Root2);

  auto Root3 = newTree({});
  std::vector<std::vector<int>> const Expected3{};
  EXPECT_EQ(Expected3, S103_1.zigzagLevelOrder(Root3));
  deleteTree(Root3);
}

TEST(BinaryTreeZigzagLevelOrderTraversal, Solution103_2) {
  auto S103_2 = Solution103_2();

  auto Root1 = newTree({3, 9, 20, NTNode, NTNode, 15, 7});
  std::vector<std::vector<int>> const Expected1{{3}, {20, 9}, {15, 7}};
  EXPECT_EQ(Expected1, S103_2.zigzagLevelOrder(Root1));
  deleteTree(Root1);

  auto Root2 = newTree({1});
  std::vector<std::vector<int>> const Expected2{{1}};
  EXPECT_EQ(Expected2, S103_2.zigzagLevelOrder(Root2));
  deleteTree(Root2);

  auto Root3 = newTree({});
  std::vector<std::vector<int>> const Expected3{};
  EXPECT_EQ(Expected3, S103_2.zigzagLevelOrder(Root3));
  deleteTree(Root3);
}

File: java/src/test/java/io/binac/leetcode/BinaryTreeZigzagLevelOrderTraversalTest.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;
import io.binac.leetcode.util.Trees;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class BinaryTreeZigzagLevelOrderTraversalTest {
    private static final BinaryTreeZigzagLevelOrderTraversal.Solution1 SOLUTION1 = new BinaryTreeZigzagLevelOrderTraversal.Solution1();
    private static final BinaryTreeZigzagLevelOrderTraversal.Solution2 SOLUTION2 = new BinaryTreeZigzagLevelOrderTraversal.Solution2();

    private final TreeNode ROOT1 = Trees.asTree(3, 9, 20, null, null, 15, 7);
    private final List<List<Integer>> EXPECTED1 = Arrays.asList(
            Collections.singletonList(3),
            Arrays.asList(20, 9),
            Arrays.asList(15, 7)
    );

    private final TreeNode ROOT2 = Trees.asTree(1);
    private final List<List<Integer>> EXPECTED2 = Collections.singletonList(Collections.singletonList(1));

    private final TreeNode ROOT3 = Trees.asTree();
    private final List<List<Integer>> EXPECTED3 = Collections.emptyList();


    @Test
    void test1() {
        assertEquals(EXPECTED1, SOLUTION1.zigzagLevelOrder(ROOT1));
        assertEquals(EXPECTED2, SOLUTION1.zigzagLevelOrder(ROOT2));
        assertEquals(EXPECTED3, SOLUTION1.zigzagLevelOrder(ROOT3));
    }

    @Test
    void test2() {
        assertEquals(EXPECTED1, SOLUTION2.zigzagLevelOrder(ROOT1));
        assertEquals(EXPECTED2, SOLUTION2.zigzagLevelOrder(ROOT2));
        assertEquals(EXPECTED3, SOLUTION2.zigzagLevelOrder(ROOT3));
    }
}